clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -0,0 +1,109 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for cache replacement policies
+#include <iostream>
+#include <assert.h>
+#include "cache_replacement_policy_unit_test.h"
+#include "simulator/cache_lru.h"
+
+class cache_lru_test_t : public cache_lru_t {
+public:
+    void
+    unit_test_replace_which_way()
+    {
+        initialize_cache(/*associativity=*/4, /*line_size=*/32, /*total_size=*/256);
+
+        // Access the cache line in the following fashion. This sequence follows the
+        // sequence shown in https://github.com/DynamoRIO/dynamorio/issues/4881.
+        const int ADDRESS_A = 0;
+        const int ADDRESS_B = 64;
+        const int ADDRESS_C = 128;
+        const int ADDRESS_D = 192;
+        const int ADDRESS_E = 72;
+
+        assert(get_block_index(ADDRESS_A) == get_block_index(ADDRESS_B));
+        assert(get_block_index(ADDRESS_B) == get_block_index(ADDRESS_C));
+        assert(get_block_index(ADDRESS_C) == get_block_index(ADDRESS_D));
+        assert(get_block_index(ADDRESS_D) == get_block_index(ADDRESS_E));
+
+        // Lower-case letter shows the least recently used way.
+        access_and_check_lru(ADDRESS_A, 1); // A X X X
+        access_and_check_lru(ADDRESS_B, 2); // A B X X
+        access_and_check_lru(ADDRESS_C, 3); // A B C X
+        access_and_check_lru(ADDRESS_D, 0); // a B C D
+        access_and_check_lru(ADDRESS_A, 1); // A b C D
+        access_and_check_lru(ADDRESS_A, 1); // A b C D
+        access_and_check_lru(ADDRESS_A, 1); // A b C D
+        access_and_check_lru(ADDRESS_E, 2); // A E c D
+
+        // XXX i#4842: Add more test sequences.
+    }
+
+private:
+    void
+    initialize_cache(int associativity, int line_size, int total_size)
+    {
+        caching_device_stats_t *stats = new cache_stats_t(line_size, "", true);
+        if (!init(associativity, line_size, total_size, nullptr, stats, nullptr)) {
+            std::cerr << "LRU cache failed to initialize"
+                      << "\n";
+            exit(1);
+        }
+    }
+
+    int
+    get_block_index(int addr)
+    {
+        addr_t tag = compute_tag(addr);
+        int block_idx = compute_block_idx(tag);
+        return block_idx;
+    }
+
+    void
+    access_and_check_lru(int addr, int way_to_replace)
```

Comment:
Make it clear that the 2nd param is *after* adding addr to the cache, and that it's just a check -- as it is doesn't it sound like this function is going to actively replace something?  So rename to `expected_replacement_way_after_access` or something.

Reply:
Done. Renamed it to `expected_replacement_way_after_access`